src/ocean/util/container/btree/Implementation.d

```diff
@@ -0,0 +1,1200 @@
+/********************************************************************************
+
+    Internal (non-user facing) implementation of BTreeMap.
+
+    Copyright:
+        Copyright (c) 2017 sociomantic labs GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.util.container.btree.Implementation;
+
+import ocean.transition;
+
+/*******************************************************************************
+
+    Internal (non-user facing) implementation of BTreeMap.
+
+    Params:
+        KeyType = type of the key
+        ValueType = type of the value
+        tree_degree = degree of the tree.
+
+*******************************************************************************/
+
+package struct BTreeMapImplementation (KeyType, ValueType, int tree_degree)
+{
+    import ocean.core.array.Mutation;
+    import ocean.core.array.Search;
+    import ocean.core.Enforce;
+    import ocean.core.Traits;
+    import ocean.core.Verify;
+    import ocean.util.container.mem.MemManager;
+
+    /**************************************************************************
+
+        Node of the tree. Contains at most (degree * 2 - 1) elements and
+        (degree * 2) subtrees.
+
+    **************************************************************************/
+
+    package struct BTreeMapNode
+    {
+        /**********************************************************************
+
+            KeyValue structure which binds key and a value to be stored in the
+            node.
+
+            NOTE: In addition of ordering notion, this is really the only thing
+            that makes this an implementation of the map, not of the set. If we're
+            going to add set support, we should probably do it via templating
+            this implementation on the actual content of the node's element
+            (replace KeyValue with just a value) and with the ordering operation
+            inserted as a policy (to compare the values, and not the nodes).
+
+        ***********************************************************************/
+
+        private struct KeyValue
+        {
+            KeyType key;
+            Unqual!(ValueType) value;
```

Comment:
Is the use of `Unqual` here described anywhere? This seems like an important implementation detail to mention at a high level.